src/typechecker/FStar.TypeChecker.TcTerm.fst

```diff
@@ -2262,10 +2262,25 @@ and tc_abs env (top:term) (bs:binders) (body:term) : term * lcomp * guard_t =
         body, cbody, Env.conj_guard guard_body g_lc
     in
 
+    if Env.debug env <| Options.Extreme
+    then BU.print1 "tc_abs: guard_body: %s\n"
+           (Rel.guard_to_string env guard_body);
+
     let guard = if env.top_level
                 || not (Options.should_verify (string_of_lid env.curmodule))
-                then Env.conj_guard (Rel.discharge_guard env g_env)
-                                    (Rel.discharge_guard envbody guard_body)
+                then begin
+                  if env.lax ||
+                     env.phase1
+                  then Env.conj_guard (Rel.discharge_guard env g_env)
+                                      (Rel.discharge_guard envbody guard_body)
+                  else
+                    let g_env, g_env_logical = TcComm.split_guard g_env in
```

Comment:
This is a minor modification to the code that was discharging `g_env` and `g_body` separately. Since the implicits in `g_env` and `g_body` may depend on each other, we get better mileage if we solve their implicits together and then discharge the two guards.

Note that the env of the two guards are different, but we can still solve their implicits together, since env doesn't matter for them.